database/scrapers/buildings.ts

```diff
@@ -0,0 +1,202 @@
+            `);
+        }
+        return data;
+      });
+    const campusesMap = new Map<string, Campus>();
+    for (const data of campusesData.data) {
+      const campus = await Campus.create(data);
+      for (const buildingIdentifier of data.buildings) {
+        campusesMap.set(buildingIdentifier, campus);
+      }
+    }
+    task.update("campuses created!");
+    task.update("starting fetching buildings...");
+    const buildingsData = await this.fetchJSON(
+      buildingsPath,
+      "list of buildings from directus",
+    );
+    if (!isValidBuildingsData(buildingsData)) {
+      throw new Error(`
+        Invalid data type fetched from ${buildingsPath}, 
+        expected type of Response<BuildingDraft>
+        `);
+    }
+    const formattedBuildingData = await Promise.all(
+      buildingsData.data.map(async (data) => {
+        const addressArray = data.addres.split(",");
+        return {
+          id: data.id,
+          identifier: data.name,
+          specialName: data.naturalName,
+          iconType: BuildingIcon.Icon,
+          addressLine1: addressArray.shift(),
+          addressLine2: addressArray.shift(),
+          latitude: data.latitude,
+          longitude: data.longitude,
+          haveFood: data.food ?? false,
+          cover: await this.semaphore.runTask(() =>
+            this.uploadCoverAndGetKey(data),
+          ),
+          externalDigitalGuideMode: data.externalDigitalGuideMode,
+          externalDigitalGuideIdOrURL: data.externalDigitalGuideIdOrURL,
+          createdAt: resolveDate(data.createdAt),
+          updatedAt: resolveDate(data.updatedAt),
+        };
+      }),
+    );
+    const updatedCampuses: Campus[] = [];
+    await Building.createMany(
+      formattedBuildingData.map((buildingEntry) => {
+        const campus = campusesMap.get(buildingEntry.identifier);
+        if (campus === undefined) {
+          throw new Error(`
+          No campus assigned to building ${buildingEntry.identifier} 
+          foreign key constraint will not be met for buildings table
+          `);
+        }
+        // swap campus cover placeholder (building identifier) for real cover key
+        if (
+          buildingEntry.cover !== undefined &&
+          campus.cover === buildingEntry.identifier
+        ) {
+          campus.cover = buildingEntry.cover;
+          updatedCampuses.push(campus);
+        }
+        return {
+          campusId: campus.id,
+          ...buildingEntry,
+        };
+      }),
+    );
+    task.update("Buildings created !");
+    //save changes for campuses
+    await Promise.all(updatedCampuses.map((campus) => campus.save()));
+    task.update("campuses cover assigned !");
+  }
+
+  private async uploadCoverAndGetKey(
+    data: BuildingDraft,
+  ): Promise<string | undefined> {
+    const imageKey = data.cover;
+    if (imageKey === null) {
+      this.logger.warning(`no image for building: [${data.id}]`);
+      return;
+    }
```

Comment:
minor nitpick: Since the result of this function is going into the database, returning `null` will work as well and should be more explicit: this way you won't be able to forget a return statement and it will be clear that you're returning a `null` value.